src/garage/sampler/base.py

```diff
@@ -1,43 +1,48 @@
+"""Base sampler class."""
+
 import abc
+import copy
+
+from garage.sampler.config import SamplerConfig
 
 
 class Sampler(abc.ABC):
-    @abc.abstractmethod
+    """Base class of all samplers."""
+
+    def __init__(self, algo, env):
+        """Construct a Sampler from an Algorithm.
+
+        Using this method is deprecated.
+        """
+        self.algo = algo
+        self.env = env
+
+    @classmethod
+    def construct(cls, config: SamplerConfig, agent, env) -> 'Sampler':
+        """Construct this sampler from a config."""
+        # This implementation works for most current implementations.
+        # Relying on this implementation is deprecated, but this method is not.
+        fake_algo = copy.copy(config)
+        fake_algo.policy = agent
+        return cls(fake_algo, env)
+
     def start_worker(self):
-        """Initialize the sampler,
+        """Initialize the sampler.
 
-        e.g. launching parallel workers if necessary.
+        i.e. launching parallel workers if necessary.
+
+        This method is deprecated, please launch workers in construct instead.
         """
-        pass
 
     @abc.abstractmethod
-    def obtain_samples(self, itr):
-        """Collect samples for the given iteration number.
-
-        Args:
-            itr (int): Iteration number.
+    def obtain_samples(self, itr, num_samples, agent_update, env_update=None):
```

Comment:
can you write in the comments what agent_update and env_update mean

Reply:
are these functions for updating the env and updating the agent? if so can we rename to agent_update_fn and env_update_fn

Reply:
These are data that will be passed into the `agent_update_fn` and `env_update_fn`. I clearly need to write more documentation for these and how they are used.

The general idea here is that we can specify one config type that works for a large number of algorithms, and then only need to specify the data to feed that config. For example, all Tensorflow policies will use the same `worker_init_fn` and `agent_update_fn`, so we can define a SamplerConfig that has the right functions for those, then specify the parameter values in each algorithm (or in one of the Algorithm base classes).

Reply:
Done.